src/compressionworker.ts

```diff
@@ -0,0 +1,97 @@
+import compress from "./compress";
+import { config } from "./config";
+
+export function createCompressionWorker(
+  sampleEnvelope: IEnvelope,
+  onMessage?: (e: MessageEvent) => void,
+  onError?: (e: ErrorEvent) => void
+): Worker {
+  let worker = null;
+  if (Worker) {
+    let workerUrl = createWorkerUrl(sampleEnvelope);
+    worker = new Worker(workerUrl);
+    worker.onmessage = onMessage || null;
+    worker.onerror = onError || null;
+  }
+  return worker;
+}
+
+function workerContext() {
+  let workerGlobalScope = self as any;
+  let compress = workerGlobalScope.compress;
+  let config = workerGlobalScope.config;
+  let envelope: IEnvelope = workerGlobalScope.envelope;
+  let nextPayload: string[] = [];
+  let nextPayloadLength = 0;
+  let sequence = 0;
+
+  // Edge case: Don't schedule next upload for XhrError instrumentation events
+  // This helps us avoid the infinite loop in the case when all requests fail (e.g. dropped internet connection)
+  // Infinite loop comes from sending instrumentation about failing to deliver previous delivery failure instrumentation.
+  let nextPayloadIsSingleXhrErrorEvent: boolean =  false;
+
+  onmessage = (evt: MessageEvent) => {
+    let message = JSON.parse(evt.data) as IWorkerMessage;
+    switch (message.type) {
+      case WorkerMessageType.AddEvent:
+        let addEventMsg = message as IAddEventMessage;
+        addEvent(addEventMsg.event, addEventMsg.time);
+        break;
+      case WorkerMessageType.ForceUpload:
+        let forceUploadMsg = message as ITimestampedWorkerMessage;
+        uploadNextPayload(forceUploadMsg.time);
+        break;
+      case WorkerMessageType.Terminate:
+        let terminatedMsg = message as ITimestampedWorkerMessage;
+        uploadNextPayload(terminatedMsg.time);
+        self.close();
+      default:
+        break;
+    }
+  };
+
+  function addEvent(event: IEvent, time: number) {
+    let eventStr = JSON.stringify(event);
+    if (nextPayloadLength > 0 && nextPayloadLength + eventStr.length > config.batchLimit) {
+      uploadNextPayload(time);
+    }
+    nextPayload.push(eventStr);
+    nextPayloadLength += eventStr.length;
+    nextPayloadIsSingleXhrErrorEvent = (nextPayload.length === 1 && event.state && event.state.type === Instrumentation.XhrError);
+    if (nextPayloadLength >= config.batchLimit) {
```

Comment:
It would, but if we are already at the limit at this point, why wait until next call ? Added some comments too

---
In reply to: [136482622](https://github.com/Microsoft/clarity-js/pull/28#discussion_r136482622) [](ancestors = 136482622)